# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:   Bryan Peralta Arboleda<br>
Url: https://github.com/A-Alexander-code/03MAIR---Algoritmos-de-Optimizacion---2023/tree/46822456af01b2ded8aa960fb797fe8083cd5a82/Proyecto%20optimizaci%C3%B3n<br>
Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>


Descripción del problema:(copiar enunciado)

Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de grabación independientemente del número de tomas que se graben. No es posible grabar más de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los
servicios de los actores de doblaje sea el menor posible.

Número de actores: 10
Número de tomas: 30
Actores/Tomas: https://bit.ly/36D8IuK

(*) La respuesta es obligatoria







(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Respuesta

In [ ]:
import numpy as np

In [ ]:
tomas = 30
sub = 6

comb = np.math.factorial(tomas)/(np.math.factorial(sub)*np.math.factorial(tomas - sub))
print(comb)

593775.0


En base a la teoría de combinaciones para una muestra de 30 tomas y una posibilidad de agruparlos en 6 conjuntos sin repetir las tomas, se tendría una cantidad de 593775 variaciones.

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Respuesta

Para reducir la cantidad de tomas, primero se busca compaginar dos tomas donde los actores no se repitan, con base en está reducción se utiliza una estructura de divide y venceras.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

La función objetivo es la compaginación de dos tomas que no repitan actores.
Este es un problema de minimización que busca reducir la cantidad de tomas y por ende el número de subconjuntos que se pueden hacer a partir de una cantidad menor de tomas.

Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

In [ ]:
# Función para agrupar las tomas en conjuntos de n integrantes

def op_schedule(rows, mod_dataSet, list_t, n):

    '''
    rows: número de tomas en el dataset
    mod_dataSet: dataset
    list_t: lista que contiene un índice de las tomas
    n: subconjuntos para agrupación
    '''

    idx_i = []

    for i in range(rows):
        if mod_dataSet[i].sum() == 0:
            idx_i.append(i)
        else:
            pass

    mod_dataSet = np.delete(mod_dataSet, idx_i, 0)

    for item in idx_i:
        list_t.remove(list_t[item])

    # Combinar tomas de tal manera que haya conjuntos de n
    new_mod = np.array(mod_dataSet, copy = True)
    n_rows, n_columns = new_mod.shape
    conj_cnd = n
    sub_c = n_rows // conj_cnd
    last_sub = n_rows - sub_c*conj_cnd
    if last_sub > 0:
        sub_c += 1
    else:
        pass

    schedule = np.zeros((sub_c, n_columns))
    j = 0
    tomas_seg = []

    for i in range(sub_c):
        schedule[i] = np.sum(new_mod[j:j+n, :], axis=0)
        j += n

    return schedule, len(list_t)

Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

La función consta de tres bucles for, dos de ellos tienen una sentencia if anidada.
<dl>Complejidad: O(n)

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

El algoritmo que se diseño aumenta su complejidad a O(n^2). Sin embargo, este algoritmo primero filtra las tomas para unir aquellas en que no hay actores repetidos y despúes agruparlas, esperando de esta manera reducir el número de subconjuntos.
<dl>
Con la implementación del algoritmo se redujo a 27 tomas de las 30 originales, lo que implica que en la última toma los gastos de movilización y necesidades de los actores son reducidos ya que en la última sesión sólo se necesita agendar y rodar con 7 participaciones, en contraste con la movilización de más de 15 participaciones en las tomas restantes.
<dl>
La estructura puede ser consultada en el archivo adjunto: optimizacion_cronograma_original.ipynb.

(*)Calcula la complejidad del algoritmo

Respuesta

Complejidad: O(n^2)

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

In [ ]:
import numpy as np
import pandas as pd
import random

random.seed(42)

def get_schedule(act, tom):

    data_set = np.zeros(shape=(tom, act))

    for i in range(tom):
        for j in range(act):
            data_set[i][j] = random.randint(0,1)

    return data_set

# Cantidad de actores y número de tomas a generarse
num_act = 12
num_tom = 28

data_set = get_schedule(num_act, num_tom)

print(data_set)

[[0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 1. 1. 0. 0. 1. 1.]
 [1. 0. 0. 1. 0. 0. 1. 0. 1. 1. 1. 0.]
 [1. 0. 1. 0. 1. 1. 0. 0. 0. 0. 1. 0.]
 [0. 0. 1. 1. 1. 1. 0. 1. 1. 0. 1. 0.]
 [0. 0. 0. 1. 1. 1. 0. 1. 0. 0. 0. 1.]
 [1. 1. 0. 0. 1. 0. 1. 1. 1. 0. 1. 0.]
 [0. 1. 1. 1. 1. 0. 0. 1. 0. 0. 0. 0.]
 [0. 1. 0. 1. 1. 1. 1. 0. 0. 1. 1. 0.]
 [1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1.]
 [0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 1. 0. 1. 1. 0. 0. 1.]
 [1. 1. 1. 1. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 1. 1. 0.]
 [0. 1. 0. 1. 1. 0. 0. 0. 1. 1. 1. 1.]
 [1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0.]
 [1. 0. 1. 1. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 1. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1.]
 [1. 1. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 1. 1. 1. 0. 1. 0. 1. 1. 0. 1.]
 [1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 1. 0.]
 [0. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 0.]
 [0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 0.]
 [1. 1. 0. 0. 1. 0. 1. 0. 1. 1. 0. 0.]
 [0. 0. 0. 1. 0. 1. 0. 0.

Aplica el algoritmo al juego de datos generado

Respuesta

El desarrollo del algoritmo puede consultarse en el archivo adjunto: optimizacion_cronograma_generado.ipynb

<dl> De las 28 tomas originales se redujo el número a 26 tomas y estas fueron agrupadas en subconjuntos de 7

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

NumPy. (s. f.). https://numpy.org/

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

Una forma de reducir aún más el número de tomas filtrado es permitiendo que en una toma que no tenga un actor que participe pueda seleccionarse otra toma, de tal manera que el número de participaciones por toma sea el máximo posible, lo que haría que el número de subconjuntos de tomas disminuya. También podría utilizarse un algoritmo de backpropagation para revisar posibles combinaciones entre tomas que ya han sido emparejadas aunque esto aumentaría la complejidad del algoritmo.